packages/smithy-client/src/client.ts

```diff
@@ -44,11 +45,21 @@ export class Client<
   ResolvedClientConfiguration extends SmithyResolvedConfiguration<HandlerOptions>,
 > implements IClient<ClientInput, ClientOutput, ResolvedClientConfiguration>
 {
-  public middlewareStack: MiddlewareStack<ClientInput, ClientOutput> = constructStack<ClientInput, ClientOutput>();
-  readonly config: ResolvedClientConfiguration;
-  constructor(config: ResolvedClientConfiguration) {
-    this.config = config;
+  public middlewareStack: MiddlewareStack<ClientInput, ClientOutput> = constructStack<ClientInput, ClientOutput>({
+    onChange: () => {
+      delete this.handlers;
+    },
+  });
+  /**
+   * May be used to cache the resolved handler function for a Command class.
+   */
+  private handlers?: WeakMap<Function, Handler<any, any>> | undefined;
+  private configRef?: ResolvedClientConfiguration | undefined;
+
+  constructor(public readonly config: ResolvedClientConfiguration) {
```

Comment:
Maybe just me not being familiar with this syntax, but it looks like `this.configRef` is what's assigned in the ctor, not `this.config`. Is `this.config` still accessible even though it isn't declared or assigned to? I.e. in the `destroy` method.

Reply:
this is shorthand, these are equivalent

```ts
public constructor(public prop: Type) {}

public constructor(prop: Type) {
  this.prop = prop;
}
```

configRef is for detecting reference changes to the client's config property, but this whole PR is kind of iffy. Probably not worth putting change detectors on so much stuff just to save 1ms of handler resolution time, and it's leaky.

Reply:
no longer applicable in latest revision